src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -79,41 +80,93 @@ internal static ValueTask<int> ReadAtOffsetAsync(SafeFileHandle handle, Memory<b
+                    // We need to try again for the remainder.
+                    buffer = buffer.Slice(bytesWritten);
+                    fileOffset += bytesWritten;
+                }
             }
         }
 
-        internal static unsafe long WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset)
+        internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnlyList<ReadOnlyMemory<byte>> buffers, long fileOffset)
         {
-            MemoryHandle[] handles = new MemoryHandle[buffers.Count];
-            Span<Interop.Sys.IOVector> vectors = buffers.Count <= IovStackThreshold ? stackalloc Interop.Sys.IOVector[IovStackThreshold] : new Interop.Sys.IOVector[buffers.Count ];
+            int buffersCount = buffers.Count;
+
+            var handles = new MemoryHandle[buffersCount];
+            Span<Interop.Sys.IOVector> vectors = buffersCount <= IovStackThreshold ?
+                stackalloc Interop.Sys.IOVector[IovStackThreshold] :
+                new Interop.Sys.IOVector[buffersCount];
 
-            long result;
             try
             {
-                int buffersCount = buffers.Count;
-                for (int i = 0; i < buffersCount; i++)
-                {
-                    ReadOnlyMemory<byte> buffer = buffers[i];
-                    MemoryHandle memoryHandle = buffer.Pin();
-                    vectors[i] = new Interop.Sys.IOVector { Base = (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
-                    handles[i] = memoryHandle;
-                }
-
-                fixed (Interop.Sys.IOVector* pinnedVectors = &MemoryMarshal.GetReference(vectors))
+                int buffersOffset = 0, firstBufferOffset = 0;
+                while (true)
                 {
-                    result = Interop.Sys.PWriteV(handle, pinnedVectors, buffers.Count, fileOffset);
+                    long totalBytesToWrite = 0;
+
+                    for (int i = buffersOffset; i < buffersCount; i++)
+                    {
+                        ReadOnlyMemory<byte> buffer = buffers[i];
+                        totalBytesToWrite += buffer.Length;
+
+                        MemoryHandle memoryHandle = buffer.Pin();
+                        vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
+                        handles[i] = memoryHandle;
+
+                        firstBufferOffset = 0;
+                    }
+
+                    long bytesWritten;
+                    fixed (Interop.Sys.IOVector* pinnedVectors = &MemoryMarshal.GetReference(vectors))
+                    {
+                        bytesWritten = Interop.Sys.PWriteV(handle, pinnedVectors, buffersCount, fileOffset);
+                    }
+
+                    FileStreamHelpers.CheckFileCall(bytesWritten, handle.Path);
+                    if (bytesWritten == totalBytesToWrite || bytesWritten == 0)
```

Comment:
same question as above

```suggestion
                    if (bytesWritten == totalBytesToWrite)
```